atef/widgets/archive_viewer.py

```diff
@@ -0,0 +1,622 @@
+                errno = e.code
+            elif e.reason.errno:
+                errno = e.reason.errno
+            else:
+                errno = -1
+
+            if 0 < errno < 400:
+                # timeouts give socket.timeout and have no errno
+                # here we only care if the server exists
+                # connection refused is ok
+                return url
+
+
+def _success_decorator(method):
+    """
+    Decorator for qt methods that want to return True if successful and
+    False otherwise.  Wraps the method in a simple try-except.
+    """
+    def _dec(self, *args, **kwargs):
+        try:
+            method(self, *args, **kwargs)
+            return True
+        except Exception as e:
+            logger.debug(e)
+            return False
+    return _dec
+
+
+class ArchiverError(Exception):
+    """ Archiver related exceptions """
+    ...
+
+
+class ArchiverViewerWidget(DesignerDisplay, QWidget):
+    """
+    Archiver time plot viewer
+    """
+    filename: ClassVar[str] = 'archive_viewer_widget.ui'
+
+    time_plot: PyDMArchiverTimePlot
+    button_month: QtWidgets.QPushButton
+    button_week: QtWidgets.QPushButton
+    button_day: QtWidgets.QPushButton
+    input_field: QtWidgets.QLineEdit
+    curve_list: QtWidgets.QTableView
+    redraw_button: QtWidgets.QPushButton
+
+    def __init__(
+        self,
+        parent: Optional[QtWidgets.QWidget] = None,
+        pvs: List[str] = []
+    ) -> None:
+        super().__init__(parent=parent)
+
+        # set the PYDM_ARCHIVER_URL if not already set
+        if not os.environ.get('PYDM_ARCHIVER_URL'):
+            archiver_url = get_reachable_url(ARCHIVER_URLS)
+            if archiver_url is None:
+                raise ArchiverError('Cannot reach any archiver urls')
+            # need to set environment variable for archiver data plugin
+            logger.debug(f'setting archiver url to: {archiver_url}')
+            # pydm requires the port to be added
+            os.environ['PYDM_ARCHIVER_URL'] = archiver_url + ':17668'
+        else:
+            archiver_url = os.environ['PYDM_ARCHIVER_URL']
+            # ensure the port has been added for pydm
+            # this handling needs work, but should suffice for now
+            if not archiver_url.endswith(':17668'):
+                port_re = re.search(r'(:\d+)$', archiver_url)
+                if port_re and (port_re[0] != ':17668'):
+                    logger.warning(f'PYDM_ARCHIVER_URL ({archiver_url}) does '
+                                   'not end with port 17668, replacing port')
+                    archiver_url = archiver_url.replace(port_re[0], ':17668')
+                else:
+                    archiver_url += ':17668'
+                os.environ['PYDM_ARCHIVER_URL'] = archiver_url
+
+        # EpicsArchive wants a stripped down url
+        url_core = archiver_url.removeprefix('http://').removesuffix(':17668')
+        self.archapp = EpicsArchive(url_core)
```

Comment:
Not sure where to put this- but it seems like it's possible to "outrun" the archiver gets by impatiently clicking buttons and cause the rendering to fail.

Reply:
```
Traceback (most recent call last):
  File "/cds/home/z/zlentz/miniconda3/envs/pcds-5.5.1/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/cds/home/z/zlentz/miniconda3/envs/pcds-5.5.1/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/cds/home/z/zlentz/github/atef/atef/__main__.py", line 3, in <module>
    main()
  File "/cds/home/z/zlentz/github/atef/atef/bin/main.py", line 99, in main
    func(**kwargs)
  File "/cds/home/z/zlentz/github/atef/atef/bin/config.py", line 46, in main
    app.exec()
  File "/cds/group/pcds/pyps/apps/dev/pythonpath/pydm/widgets/archiver_time_plot.py", line 112, in receiveArchiveData
    self.getViewBox().addItem(self.error_bar_item)
AttributeError: 'NoneType' object has no attribute 'addItem'
```

Reply:
I was always able to reset and get something after clicking more patiently. I'm not sure if anything can be done to make it obvious that the gui is waiting for an archiver response.

Reply:
I'm starting to suspect this to be more of a pydm side bug than something I can fix here...  This can more easily be replicated if you load a few PV's and increase the range to 1 month or something large.  Before all three can grab archive data, the plot will complain.  The data loads just fine afterwards though

Reply:
It's absolutely possible that it's a pydm bug